### PR TITLE
Removal of all compiler warnings for build on CentOS 6

### PR DESCRIPTION
**Pull request by fquinner**
This set of changes removes all warnings from our CI environment.

Noteably it also:

- Removes the legacy createImpl methods from qpid and avis
- Fixes bug in deprecation warning for gcc 5+
- Adds macros to replace enum values (e.g. MAMA_PAYLOAD_ID_QPID)
- Removes deprecated unit tests which depend on old enum behaviour
- Removes dead code

Signed-off-by: Frank Quinn <fquinn.ni@gmail.com>

[Review on Reviewable](https://reviewable.io/reviews/openmama/openmama/142)

### CODE REVIEW

**Comment by fquinner** [comment]
CI branch for this: http://ci.openmama.org:8080/job/feature-warning-cull/

File: common/c_cpp/src/c/linux/port.h [hunk 1]
```diff
@@ -197,7 +197,7 @@ const char* getHostName (void);
 #      define MAMADeprecated__(MSG)     __attribute__ ((deprecated(#MSG)))
 #    endif
 /* GCC 4.5 and greater also supports message with deprecated */
-#  elif (__GNUC__ >= 4 && __GNUC_MINOR__ >= 5)
+#  elif (__GNUC__ >= 4 && __GNUC_MINOR__ >= 5) || __GNUC__ > 4
```
Comment (MattMulhern) [inline common/c_cpp/src/c/linux/port.h]: Does this not start to include GCC versions 4 - 4.4 inclusive now? 
If i recall correctly (and judging my the comment above) the __attribute__ calls may only be compatible on gcc 4.5 onwards?

File: mama/c_cpp/src/c/SConscript [hunk 2]
```diff
@@ -183,18 +183,26 @@ else:
...
+shared_objects = []
 StaticLibs['libmama'] = []
 
 
 for s in libmama_sources:
-    obj = env.Object(s)
-    StaticLibs['libmama'].append(obj)
-    objects.append(obj)
+    if s == "msg.c":
+        static_obj = env.Object(s, CFLAGS="-Wno-deprecated-declarations")
```
Comment (MattMulhern) [inline mama/c_cpp/src/c/SConscript]: can we stick a comment in here explaining why we can't use our "ignore deprecated" macros?

File: mama/c_cpp/src/c/bridge/avis/publisher.c [hunk 3]
```diff
@@ -100,7 +101,9 @@ avisBridgeMamaPublisherImpl_prepareMessage (mamaMsg* msg)
         return ret;
     }
 
-    if (MAMA_PAYLOAD_AVIS == payload)
+    // TODO: Should replace this with a more run-time way to get the payload ID
```
Comment (MattMulhern) [inline mama/c_cpp/src/c/bridge/avis/publisher.c]: Should we spin out an issue for covering this in some later release? Just so we don't commit a TODO and never address it.

File: mama/c_cpp/src/c/mama.c [hunk 4]
```diff
@@ -235,6 +242,16 @@ mama_loadPayloadBridgeInternal  (mamaPayloadBridge* impl,
 mama_status
 mama_loadEntitlementBridgeInternal  (const char* name);
 
+void MAMACALLTYPE mamaImpl_entitlementDisconnectCallback (
```
Comment (MattMulhern) [inline mama/c_cpp/src/c/mama.c]: should we move this to `entitlementinternal.h` (redefined identically in mama.c)?

File: mama/c_cpp/src/cpp/MamaSymbolList.cpp [hunk 5]
```diff
@@ -162,6 +163,7 @@ namespace Wombat
 
             return MAMA_STATUS_OK;
         }
+        */
```
Comment (MattMulhern) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: If this is ending up as dead code, should we not just delete it?

File: mama/c_cpp/src/cpp/MamaSymbolList.cpp [hunk 6]
```diff
@@ -177,6 +179,7 @@ namespace Wombat
         mamaSymbolList_setClosure             (myList, myPimpl);
         mamaSymbolList_setAddSymbolHandler    (myList, addMemberHandler);
         mamaSymbolList_setRemoveSymbolHandler (myList, removeMemberHandler);
+        // TODO: See why this is disabled
```
Comment (MattMulhern) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: Again, spin out an issue for investigation if we're not covering it in this release?

File: site_scons/community/linux.py [hunk 7]
```diff
@@ -104,9 +104,11 @@ def create( self, opts ):
     #
     # Configures all the necessary environment variables for Linux
     def configure(self, env ):
-        env.Append( CPPDEFINES = ['HAVE_CONFIG_H','_GNU_SOURCE'] )
+        env.Append( CPPDEFINES = ['_GNU_SOURCE'] )
```
Comment (MattMulhern) [inline site_scons/community/linux.py]: Can we remove the duplicate commented out lines please?

**Comment by MattMulhern** [comment]
Just a few things I'm unsure about commented on here, but apart from these it looks great!

File: mama/c_cpp/src/cpp/MamaSymbolList.cpp (hunk 5 above)
Comment (fquinner) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: Wanted to leave it in as wasn't sure if this was  a bug or not - it's part of the same change as the one below. Will raise a new ticket to track.

File: mama/c_cpp/src/cpp/MamaSymbolList.cpp (hunk 6 above)
Comment (fquinner) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: As above

File: site_scons/community/linux.py (hunk 7 above)
Comment (fquinner) [inline site_scons/community/linux.py]: I actually left those in on purpose to make it easy to switch between making warnings fatal or not, though I can take them out.

File: mama/c_cpp/src/c/SConscript (hunk 2 above)
Comment (fquinner) [inline mama/c_cpp/src/c/SConscript]: On its way

**Comment by fquinner** [comment]
I actually forgot about the payload ID thing - will think of a better way to fix this and revert (I don't want any dependency on that ID where I can avoid it). Let's chat about some of the others tomorrow.

File: common/c_cpp/src/c/linux/port.h (hunk 1 above)
Comment (fquinner) [inline common/c_cpp/src/c/linux/port.h]: Will change to >= 5 for readability.

File: mama/c_cpp/src/c/mama.c (hunk 4 above)
Comment (fquinner) [inline mama/c_cpp/src/c/mama.c]: As discussed, will rename these to avoid callback function name conflicts just for the sake of clarity.

Comment (fquinner) [inline mama/c_cpp/src/c/mama.c]: Actually I stand corrected on closer inspection it seems this function is only *called* in oea.c, so going to leave this as-is.

File: mama/c_cpp/src/c/mama/msg.h [hunk 8]
```diff
@@ -36,8 +36,26 @@
...
+#define MAMA_PAYLOAD_ID_UMS             'L'
+#define MAMA_PAYLOAD_ID_INRUSH          'N'
+#define MAMA_PAYLOAD_ID_TICK42RMDS      'P'
+#define MAMA_PAYLOAD_ID_QPID            'Q'
+#define MAMA_PAYLOAD_ID_TIBRV           'R'
+#define MAMA_PAYLOAD_ID_IBMWFO          'S'
+#define MAMA_PAYLOAD_ID_ACTIV           'T'
+#define MAMA_PAYLOAD_ID_VULCAN          'V'
+#define MAMA_PAYLOAD_ID_WOMBAT_MSG      'W'
+#define MAMA_PAYLOAD_ID_EXEGY           'X'
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/msg.h]: Actually going to remove all these again - the more I think about it the more I think we should avoid these IDs living in MAMA code entirely.

File: mama/c_cpp/src/c/bridge/avis/publisher.c (hunk 3 above)
Comment (fquinner) [inline mama/c_cpp/src/c/bridge/avis/publisher.c]: Raised https://github.com/OpenMAMA/OpenMAMA/issues/143, will comment out TODO.

File: mama/c_cpp/src/cpp/MamaSymbolList.cpp (hunk 5 above)
Comment (fquinner) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: Raised https://github.com/OpenMAMA/OpenMAMA/issues/143, removing TODO in code.

Comment (fquinner) [inline mama/c_cpp/src/cpp/MamaSymbolList.cpp]: Done

File: site_scons/community/linux.py (hunk 7 above)
Comment (fquinner) [inline site_scons/community/linux.py]: Done

**Comment by MattMulhern** [comment]
pushed up to next (6236c21fd26a49df97673b546321d4a71197bf28).

**Pull request closed without merging**